ted2zim/scraper.py

```diff
@@ -787,105 +810,130 @@ def download_thumbnail(
-            self.download_speaker_image(
-                video_id, video_title, video_speaker, speaker_path
+        # set up variables
+        video_id = str(video["id"])
+        # Take the english version of title or else whatever language it's available in
+        video_title = video["title"][0]["text"]
+        video_link = video["video_link"]
+        video_speaker = video["speaker_picture"]
+        video_thumbnail = video["thumbnail"]
+        video_dir = self.videos_dir.joinpath(video_id)
+        org_video_file_path = video_dir.joinpath("video.mp4")
+        req_video_file_path = video_dir.joinpath(f"video.{self.video_format}")
+        speaker_path = video_dir.joinpath("speaker.webp")
+        thumbnail_path = video_dir.joinpath("thumbnail.webp")
+
+        # ensure that video directory exists
+        if not video_dir.exists():
+            video_dir.mkdir(parents=True)
+
+        # set preset
+        preset = {"mp4": VideoMp4Low}.get(self.video_format, VideoWebmLow)()
+
+        # download video
+        downloaded_from_cache = False
+        logger.debug(f"Downloading {video_title}")
+        if self.s3_storage:
+            s3_key = f"{self.video_format}/{self.video_quality}/{video_id}"
+            downloaded_from_cache = self.download_from_cache(
+                s3_key, req_video_file_path, preset.VERSION
             )
-            self.download_thumbnail(
-                video_id, video_title, video_thumbnail, thumbnail_path
+        if not downloaded_from_cache:
+            try:
+                if "https://" not in video_link:
+                    self.yt_downloader.download(video_link, org_video_file_path)
+                else:
+                    save_large_file(video_link, org_video_file_path)
+            except Exception:
+                logger.error(f"Could not download {org_video_file_path}")
+
+        # download speaker and thumbnail images
+        self.download_speaker_image(video_id, video_title, video_speaker, speaker_path)
+        self.download_thumbnail(video_id, video_title, video_thumbnail, thumbnail_path)
+
+        # recompress if necessary
+        try:
+            post_process_video(
+                video_dir,
+                video_id,
+                preset,
+                self.video_format,
+                self.low_quality,
+                downloaded_from_cache,
             )
+        except Exception as e:
+            logger.error(f"Failed to post process video {video_id}")
+            logger.debug(e)
+        else:
+            # upload to cache only if recompress was successful
+            if self.s3_storage and not downloaded_from_cache:
+                self.upload_to_cache(s3_key, req_video_file_path, preset.VERSION)
+
+    def download_video_files_parallel(self):
+        """ download videos and images parallely """
+
+        self.yt_downloader = YoutubeDownloader(threads=1 if self.threads == 1 else 2)
+        with concurrent.futures.ThreadPoolExecutor(
```

Comment:
This is because of the following reasons -
1. All videos are not hosted on YouTube. The scraperlib thing only parallelizes (or throttles) the youtube downloading
2. It does everything related to videos - Downloading videos, thumbnails and speaker images, and upload to cache

Reply:
OK